bevy_lint/src/sym.rs

```diff
@@ -0,0 +1,186 @@
+//!         </tr>
+//!         <tr>
+//!             <td>...</td>
+//!         </tr>
+//!         <tr>
+//!             <td><code>SYMBOL_OFFSET</code></td>
+//!             <td rowspan="2">Symbols in <code>bevy_lint::sym</code></td>
+//!         </tr>
+//!         <tr>
+//!             <td>...</td>
+//!         </tr>
+//!     </tbody>
+//! </table>
+//!
+//! Note that the order here is important. [`clippy_utils`] expects its symbols to start at
+//! [`PREDEFINED_SYMBOLS_COUNT`], which is why it's before Bevy's symbols.
+
+#![allow(
+    non_upper_case_globals,
+    reason = "Symbol constants are named as-is so it is easy to see what strings they represent."
+)]
+
+use clippy_utils::sym::EXTRA_SYMBOLS as CLIPPY_SYMBOLS;
+/// These are symbols that we use but are already interned by either the compiler or Clippy.
+pub use rustc_span::sym::{bevy_ecs, plugin, reflect};
+use rustc_span::{Symbol, symbol::PREDEFINED_SYMBOLS_COUNT};
+
+/// The starting offset used for the first Bevy-specific symbol.
+///
+/// This is used instead of [`PREDEFINED_SYMBOLS_COUNT`] because this takes into account Clippy's
+/// pre-interned symbols as well.
+const SYMBOL_OFFSET: u32 = PREDEFINED_SYMBOLS_COUNT + CLIPPY_SYMBOLS.len() as u32;
+
+/// A helper used by `declare_bevy_symbols!` to extract its input.
+///
+/// ```ignore
+/// assert_eq!(extract_value!(Name), "Name");
+/// assert_eq!(extract_value!(Name: "value"), "value");
+/// ```
+macro_rules! extract_value {
+    ($name:ident) => {
+        stringify!($name)
+    };
+    ($name:ident: $value:literal) => {
+        $value
+    };
+}
+
+/// Generates the [`Symbol`] constants and [`BEVY_SYMBOLS`] from a list of name-value pairs.
+///
+/// # Example
+///
+/// ```ignore
+/// declare_bevy_symbols! {
+///     // Interns the string "Hello, world" available as the constant named `Hello`.
+///     Hello: "Hello, world!",
+///     // Interns the string "bevy" available as the constant named `bevy`. This is the shorthand!
+///     bevy,
+/// }
+/// ```
+macro_rules! declare_bevy_symbols {
+    {
+        $($name:ident $(: $value:literal)?),* $(,)?
+    } => {
+        /// A list of strings that are pre-interned at the beginning of linting through
+        /// [`Config::extra_symbols`](rustc_interface::interface::Config::extra_symbols).
+        const BEVY_SYMBOLS: &[&str] = &[
+            $(
+                extract_value!($name $(: $value)?)
+            ),*
+        ];
+
+        $(
+            #[doc = concat!("A pre-interned [`Symbol`] for the string \"", extract_value!($name $(: $value)?), "\".")]
+            pub const $name: Symbol = Symbol::new(SYMBOL_OFFSET + ${index()});
+        )*
+    };
+}
+
+// Before adding a new symbol here, check that it doesn't exist yet in `rustc_span::sym` or
+// `clippy_utils::sym`. Having duplicate symbols will cause the compiler to ICE! Also please keep
+// this list alphabetically sorted :)
+declare_bevy_symbols! {
+    add_systems,
+    app,
+    App,
```

Comment:
This will result in the constants:

```rust
pub const add_systems: Symbol = Symbol::new(SYMBOL_OFFSET + 0);
pub const app: Symbol = Symbol::new(SYMBOL_OFFSET + 1);
pub const App: Symbol = Symbol::new(SYMBOL_OFFSET + 2);
// ...
```